### PR TITLE
detach spectral norm calculated weight in eval mode

### PR DESCRIPTION
**Pull request by t-vi**
As we left weight to be the last calculated weight in eval mode, we need to detach it from the computation in order to facilitate using backward.
The typical use case is in GANs when the discriminator has spectral norm, is in eval mode and we want to backprop through the discriminator to get weight gradients for the generator.

### CODE REVIEW

**Comment by t-vi** [comment]
@SsnL so this was missing, too.

**Comment by t-vi** [comment]
One of the builds seem to have hit a timeout in a caffe2 test. I'm not sure that I can fix this by modifying the PR.

**Comment by ezyang** [comment]
It's not your problem.

**Review by ezyang — APPROVED** [review]

File: torch/nn/utils/spectral_norm.py [hunk 1]
```diff
@@ -51,6 +51,8 @@ def __call__(self, module, inputs):
             weight, u = self.compute_weight(module)
             setattr(module, self.name, weight)
             setattr(module, self.name + '_u', u)
+        else:
+            getattr(module, self.name).detach_()
```
Comment (SsnL) [inline torch/nn/utils/spectral_norm.py]: This should have `.requires_grad_()`. Arguably the weight should still be backprop-able in eval mode.

**Review by ssnl — CHANGES_REQUESTED** [review]

File: torch/nn/utils/spectral_norm.py (hunk 1 above)
Comment (t-vi) [reply torch/nn/utils/spectral_norm.py]: But is it OK to reset this for each iteration or should I set to the previous grad (or that of weight_orig)?

Comment (SsnL) [reply torch/nn/utils/spectral_norm.py]: Set previous `requires_grad` is definitely better! Go for it.

Comment (SsnL) [reply torch/nn/utils/spectral_norm.py]: `weight`'s `requires_grad` should be the same as that of `weight_orig`, right?

Comment (t-vi) [reply torch/nn/utils/spectral_norm.py]: Well, only `weight_orig` shows up in the parameters, so this situation might not be too uncommon:
I'm looking at this from the perspective of my (possibly not terribly good) [SN-GAN implementation](https://nbviewer.jupyter.org/github/t-vi/pytorch-tvmisc/blob/master/wasserstein-distance/sn_projection_cgan_64x64_1000c.ipynb)
For the generator update I do
```
discriminator.eval()
for p in discriminator.parameters():
   p.requires_grad = False
```
this sees weight_orig, but not weight. So weight_orig will be `requires_grad = False` and weight `requires_grad = True`. One might argue that setting weight's requires_grad to weight_orig's is less surprising here.

Comment (t-vi) [reply torch/nn/utils/spectral_norm.py]: So I changed it to weight_orig's requires grad, I hope that's OK.

Comment (SsnL) [reply torch/nn/utils/spectral_norm.py]: Yeah it should be okay. My argument was that since `weight` is computed from `weight_orig` (in most cases), it should have the same `requires_grad` flag.

**Review by ssnl — APPROVED** [review]
